server.go

```diff
@@ -3575,6 +3575,81 @@ func (s *server) peerTerminationWatcher(p *peer.Brontide, ready chan struct{}) {
 				"connection to peer %v", p)
 
 			s.connMgr.Connect(connReq)
+
+			var connected bool
+			updateConnReq := func() bool {
+				s.mu.Lock()
+				defer s.mu.Unlock()
+
+				// Check if there are any existing connection
+				// requests for this peer. If there are not
+				// then we have already successfully connected
+				// to them.
+				pubStr := string(pubKey.SerializeCompressed())
+				_, ok := s.persistentConnReqs[pubStr]
+				if !ok {
+					connected = true
+					return false
+				}
+
+				backoff = s.nextPeerBackoff(
+					pubStr, p.StartTime(),
+				)
+				s.persistentPeersBackoff[pubStr] = backoff
+
+				// Refresh the peers latest advertised address
+				// from the DB.
+				advertisedAddr, err := s.fetchNodeAdvertisedAddr(
+					pubKey,
+				)
+				if err != nil {
+					srvrLog.Errorf("Unable to "+
+						"retrieve advertised "+
+						"address for node %x: %v",
+						p.PubKey(), err)
+
+					return false
+				}
+
+				// If the existing connection request uses
+				// the latest advertised address then there
+				// is no need to update the connection request.
+				if advertisedAddr == connReq.Addr {
+					return false
+				}
+
+				// We update the connection request with the
+				// newly advertised address.
+				p.SetAddress(advertisedAddr)
+				connReq.Addr = p.NetAddress()
+
+				s.persistentConnReqs[pubStr] = append(
+					s.persistentConnReqs[pubStr], connReq,
+				)
+
+				return true
+			}
+
+			for !connected {
+				select {
```

Comment:
I think the `select` statement from L3566 to L3577 can be removed? It seems you already have a nice retry func here, no need to do that select above.

Reply:
the reason i left that select  in is that the logic is different depending on if it is the first time `s.connMgr.Connect(connReq)` that vs subsequent times. For the first time, we definitely want to call `s.connMgr.Connect(connReq)`. The subsequent times, we only call that if connReq address has changed

Reply:
but could also add a `firstReconnect` bool if you think that would be better?

Reply:
Was thinking something like this, easier to show in code,

```go
        // First, cleanup any remaining state the server has regarding the peer
	// in question.
	s.removePeer(p)

	// Next, check to see if this is a persistent peer or not.
	_, ok := s.persistentPeers[pubStr]
	if !ok {
		return
	}

	// We'll only need to re-launch a connection request if one
	// isn't already currently pending.
	if _, ok := s.persistentConnReqs[pubStr]; ok {
		return
	}

	go func() {
		for {
			// We'll ensure that we locate an advertised address to use
			// within the peer's address for reconnection purposes.
			//
			// TODO(roasbeef): use them all?
			advertisedAddr, err := s.fetchNodeAdvertisedAddr(pubKey)
			switch {
			// We found an advertised address, so use it.
			case err == nil:
				p.SetAddress(advertisedAddr)

			// The peer doesn't have an advertised address.
			case err == errNoAdvertisedAddr:
				// Fall back to the existing peer address if
				// we're not accepting connections over Tor.
				if s.torController == nil {
					break
				}

				// If we are, the peer's address won't be known
				// to us (we'll see a private address, which is
				// the address used by our onion service to dial
				// to lnd), so we don't have enough information
				// to attempt a reconnect.
				srvrLog.Debugf("Ignoring reconnection attempt "+
					"to inbound peer %v without "+
					"advertised address", p)
				return

			// We came across an error retrieving an advertised
			// address, log it, and fall back to the existing peer
			// address.
			default:
				srvrLog.Errorf("Unable to retrieve advertised "+
					"address for node %x: %v", p.PubKey(),
					err)
			}

			// Otherwise, we'll launch a new connection request in order to
			// attempt to maintain a persistent connection with this peer.
			connReq := &connmgr.ConnReq{
				Addr:      p.NetAddress(),
				Permanent: true,
			}
			s.persistentConnReqs[pubStr] = append(
				s.persistentConnReqs[pubStr], connReq)

			// Record the computed backoff in the backoff map.
			backoff := s.nextPeerBackoff(pubStr, p.StartTime())
			s.persistentPeersBackoff[pubStr] = backoff

			// Initialize a retry canceller for this peer if one does not
			// exist.
			cancelChan, ok := s.persistentRetryCancels[pubStr]
			if !ok {
				cancelChan = make(chan struct{})
				s.persistentRetryCancels[pubStr] = cancelChan
			}

			// We choose not to wait group this go routine since the Connect
			// call can stall for arbitrarily long if we shutdown while an
			// outbound connection attempt is being made.
			srvrLog.Debugf("Scheduling connection re-establishment to "+
				"persistent peer %v in %s", p, backoff)

			select {
			case <-time.After(backoff):
			// If connected, the cancelChan will be closed.
			case <-cancelChan:
				return
			case <-s.quit:
				return
			}

			srvrLog.Debugf("Attempting to re-establish persistent "+
				"connection to peer %v", p)

			s.connMgr.Connect(connReq)
		}
	}()
}
```

Reply:
Further looked into the mechanism, things I'm not sure,
- when connected,  we will close the `cancelChan`? Thus we could rely on that instead of a `connected` bool.
- I noticed there's a lock used in `updateConnReq`. This part I'm not sure either, as there's a lock used above at L3453. Will this create a deadlock?